File: adapters/python/agentgit_adapter/guards/registry.py
"""Run a chain of guards in registration order.

Stops at the first blocking guard. Snapshot hashes from passing guards
are surfaced on the final result (last one wins, mirroring TS).
"""

from __future__ import annotations

from typing import Any, Dict, List, Sequence


class GuardRegistry:
    def __init__(self, guards: Sequence[Any]) -> None:
        self._guards = list(guards)

    @property
    def size(self) -> int:
        return len(self._guards)

    def run(self, tool_call: Dict[str, Any]) -> Dict[str, Any]:
        last_snapshot: Any = None
        for guard in self._guards:
            result = guard.check(tool_call)
            if "snapshot_hash" in result:
                last_snapshot = result["snapshot_hash"]
            if result.get("outcome") == "block":
                return result
        out: Dict[str, Any] = {"outcome": "allow"}
        if last_snapshot is not None:
            out["snapshot_hash"] = last_snapshot
        return out

File: packages/core/src/guards/registry.ts
import { performance } from "node:perf_hooks";
import type { ObjectStore } from "../object-store.js";
import { safeRecord, type Reporter } from "../telemetry/reporter.js";
import type { ToolCall } from "../types.js";
import type { Guard, GuardContext, GuardResult } from "./types.js";

/**
 * Runs a chain of guards against a tool call in registration order.
 * Stops at the first blocking guard; all snapshot hashes from passing guards
 * are surfaced in the final result.
 *
 * @public
 */
export class GuardRegistry {
  /**
   * @param guards - chain to evaluate, in order.
   * @param reporter - optional telemetry sink; emits one `guard.evaluate` span
   *   per `runGuards` call when present.
   */
  constructor(
    private readonly guards: Guard[],
    private readonly reporter: Reporter | null = null,
  ) {}

  async runGuards(
    toolCall: ToolCall,
    objectStore?: ObjectStore,
  ): Promise<GuardResult> {
    const t0 = performance.now();
    const context: GuardContext =
      objectStore !== undefined ? { toolCall, objectStore } : { toolCall };
    const snapshotHashes: string[] = [];

    let blocked = false;
    let final: GuardResult = { outcome: "allow" };
    for (const guard of this.guards) {
      const result = await guard.check(context);

      if (result.snapshotHash !== undefined) {
        snapshotHashes.push(result.snapshotHash);
      }

      if (result.outcome === "block") {
        blocked = true;
        final = result;
        break;
      }
    }

    if (!blocked) {
      const lastHash = snapshotHashes[snapshotHashes.length - 1];
      if (lastHash !== undefined) {
        final.snapshotHash = lastHash;
      }
    }

    safeRecord(this.reporter, {
      name: "guard.evaluate",
      durationMs: performance.now() - t0,
      // Privacy: no tool name, no tool input/output, no snapshot hash.
      // outcome is "allow"|"block" — a categorical aggregate, not user data.
      attrs: {
        outcome: final.outcome,
        guards: this.guards.length,
      },
    });

    return final;
  }

  get size(): number {
    return this.guards.length;
  }
}
